rust/src/storage/s3/dynamodb.rs

```diff
@@ -0,0 +1,311 @@
+    pub const IS_RELEASED_PATH: &str = "#ir";
+    pub const IS_RELEASED_VALUE: &str = ":ir";
+}
+
+pub struct DynamoLock {
+    client: DynamoDbClient,
+    table_name: String,
+}
+
+#[derive(Clone, Debug)]
+pub struct LockItem {
+    owner_name: String,
+    record_version_number: String,
+    lease_duration: u64,
+    is_released: bool,
+
+    lookup_time: u128,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum DynamoError {
+    #[error("Dynamo table not found")]
+    TableNotFound,
+
+    #[error("Condition failed: {0}")]
+    ConditionalCheckFailed(String),
+
+    #[error("DynamoDB item has invalid schema")]
+    InvalidItemSchema,
+
+    #[error("Could not acquire lock for {0} sec")]
+    TimedOut(u64),
+
+    #[error("Get item error: {0}")]
+    GetItemError(RusotoError<GetItemError>),
+
+    #[error("Put item error: {0}")]
+    PutItemError(RusotoError<PutItemError>),
+}
+
+impl DynamoLock {
+    pub fn new() -> Self {
+        let client = DynamoDbClient::new(Region::default());
+        Self {
+            client,
+            table_name: String::from("mykhailo_test_table_2"),
+        }
+    }
+
+    async fn acquire_lock(&self) -> Result<LockItem, DynamoError> {
+        let mut lock_trying_to_be_acquired: Option<LockItem> = None;
+        let started = Instant::now();
+        let mut dur_to_wait = Duration::from_secs(1);
+
+        println!("Start {:?}", started);
+        while started.elapsed() < dur_to_wait {
+            let existing_lock = self.get_lock_item().await?;
+            let new_lock = LockItem::new();
+
+            let call = match existing_lock {
+                None => {
+                    println!("upsert_new_lock");
+                    self.upsert_new_lock(&new_lock).await
+                }
+                Some(ref l) if l.is_released => {
+                    println!("upsert_released_lock");
+                    self.upsert_released_lock(&new_lock).await
+                }
+                Some(ref existing) => match lock_trying_to_be_acquired {
+                    None => {
+                        println!("lock_trying_to_be_acquired None");
+                        lock_trying_to_be_acquired = existing_lock.clone();
+                        dur_to_wait = Duration::from_secs(5 + existing.lease_duration);
+                        std::thread::sleep(Duration::from_secs(1));
+                        continue;
+                    }
+                    Some(ref prev)
+                        if existing.record_version_number == prev.record_version_number =>
+                    {
+                        if prev.is_expired() {
+                            println!("lock_trying_to_be_acquired Some/eq/is_expired");
+                            self.upsert_expired_lock(prev, &new_lock).await
+                        } else {
+                            println!("lock_trying_to_be_acquired Some/eq/");
+                            std::thread::sleep(Duration::from_secs(1));
```

Comment:
should use async sleep like https://docs.rs/tokio/1.4.0/tokio/time/fn.sleep.html in async method